_LowLevelCode/cpp/GetMD5/GetMD5.c

```diff
@@ -0,0 +1,919 @@
+}
+
+// Elementary array as byte stream: ============================================
+void ProcessBin(uchar_T *array, mwSize inputLen, uchar_T digest[16])
+{
+  // Only the contents of the array is considered. Therefore double(0) and
+  // single([0,0]) reply the same hash. This works for numeric, char and logical
+  // arrays only, neitehr cells nor structs.
+  MD5_CTX context;
+  
+  MD5_Init(&context);
+  MD5_Update(&context, array, inputLen);
+  MD5_Final(digest, &context);
+}
+
+// File as byte stream: ========================================================
+void ProcessFile(char *filename, uchar_T digest[16])
+{
+  FILE    *FID;
+  MD5_CTX context;
+  mwSize  len;
+  
+  // Open the file in binary mode:
+  if ((FID = fopen(filename, "rb")) == NULL) {
+     ERROR3("MissFile", "Cannot open file: [%s]", filename);
+  }
+  
+  MD5_Init(&context);
+  while ((len = fread(buffer, 1, BUFFER_LEN, FID)) != 0) {
+     MD5_Update(&context, buffer, len);
+  }
+  MD5_Final(digest, &context);
+
+  fclose(FID);
+}
+
+// Output of 16 uchar_Ts as 32 character hexadecimals: ===========================
+void ToHex(const uchar_T digest[16], char *output, int LowerCase)
+{
+  char *outputEnd, *Fmt;
+  const uchar_T *s = digest;
+  
+  Fmt = LowerCase ? "%02x" : "%02X";
+  
+  for (outputEnd = output + 32; output < outputEnd; output += 2) {
+     sprintf(output, Fmt, *(s++));
+  }
+}
+
+// BASE64 encoded output: ======================================================
+void ToBase64(const uchar_T In[16], char *Out)
+{
+   // The base64 encoded string is shorter than the hex string.
+   // Needed length: ((len + 2) / 3 * 4) + 1, here fixed to 22+1 (trailing 0!).
+   static const uchar_T B64[] =
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+   int   i;
+   char  *p;
+   const uchar_T *s;
+   
+   p = Out;
+   s = In;
+   for (i = 0; i < 5; i++) {
+      *p++ = B64[(*s >> 2) & 0x3F];
+      *p++ = B64[((*s & 0x3) << 4)   | ((s[1] & 0xF0) >> 4)];
+      *p++ = B64[((s[1] & 0xF) << 2) | ((s[2] & 0xC0) >> 6)];
+      *p++ = B64[s[2] & 0x3F];
+      s   += 3;
+   }
+   
+   *p++ = B64[(*s >> 2) & 0x3F];
+   *p++ = B64[((*s & 0x3) << 4)];
+   *p   = '\0';
+}
+
+// Main function: ==============================================================
+void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
+{
+  // Mex interface:
+  // - Define default values of optional arguments.
+  // - Forward input data to different calculators according to the input type.
+  // - Convert digest to output format.
+  
+  char     *FileName, outString[33];
+  uchar_T  digest[16], *digestP, OutType = 'h';
+  double   *outP, *outEnd;
+  Method_t Method = BINARY_m;
+  mwSize   nByte;
+  
+  // Check number of inputs and outputs: ---------------------------------------
+  if (nrhs == 0 || nrhs > 3) {
+     ERROR("BadNInput", "1 to 3 inputs required.");
+  }
+  if (nlhs > 1) {
+     ERROR("BadNOutput", "Too many output arguments.");
+  }
```

Comment:
All our mex files contain standard output, returned when mex is called with no inputs and single output. The `mex`-es all included in `check_horace_mex` which is called at initialization and verifies if mex files are available and compiled correctly. If they are not, mex gets disabled. 

I think this mex should follow standard practice.

Reply:
thanks, should be done